### PR TITLE
Adapt rotate/rotate_copy to C++20

### PR DESCRIPTION
**Pull request by hcq9102**
adapt rotate and rotate_copy to c++20,
add container overloads(range and sentinel),
add container tests.

## Any background context you want to provide?
Issue #4822

### CODE REVIEW

**Comment by hkaiser** [comment]
add to whitelist

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp [hunk 1]
```diff
@@ -97,55 +264,16 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
-    ///           \a hpx::future<tagged_pair<tag::begin(FwdIter), tag::end(FwdIter)> >
-    ///           if the execution policy is of type
-    ///           \a parallel_task_policy and
-    ///           returns \a tagged_pair<tag::begin(FwdIter), tag::end(FwdIter)>
-    ///           otherwise.
-    ///           The \a rotate algorithm returns the iterator equal to
-    ///           pair(first + (last - new_first), last).
-    ///
+
+    //clang-format off
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp:268]: nitpick: could you write this as `// clang-format off` (i.e. add a space after the `//`)

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp [hunk 2]
```diff
@@ -97,55 +264,16 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+    //clang-format off    
     template <typename ExPolicy, typename FwdIter,
-        HPX_CONCEPT_REQUIRES_(hpx::is_execution_policy<ExPolicy>::value&&
-                hpx::traits::is_iterator<FwdIter>::value)>
+        HPX_CONCEPT_REQUIRES_(
+            hpx::is_execution_policy<ExPolicy>::value&&
+            hpx::traits::is_iterator<FwdIter>::value
+        )>
+    //clang-format on 
+    HPX_DEPRECATED_V(1, 7,
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp:275]:
```suggestion
    HPX_DEPRECATED_V(1, 8,
```

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp [hunk 3]
```diff
@@ -308,3 +395,120 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+            )>
+        // clang-format on
+        friend FwdIter tag_fallback_dispatch(
+            hpx::rotate_t, FwdIter first, FwdIter new_first, FwdIter last)
+        {
+            static_assert(hpx::traits::is_forward_iterator<FwdIter>::value,
+                "Requires at least forward iterator.");
+
+            return hpx::parallel::v1::detail::rotate<FwdIter>().call2(
+                hpx::execution::sequenced_policy{}, first, new_first, last);
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp:420]:
```suggestion
                hpx::execution::seq, first, new_first, last);
```

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp [hunk 4]
```diff
@@ -308,3 +395,120 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+                hpx::traits::is_forward_iterator<FwdIter>::value
+            )>
+        // clang-format on
+        friend FwdIter tag_fallback_dispatch(
+            hpx::rotate_t, FwdIter first, FwdIter new_first, FwdIter last)
+        {
+            static_assert(hpx::traits::is_forward_iterator<FwdIter>::value,
+                "Requires at least forward iterator.");
+
+            return hpx::parallel::v1::detail::rotate<FwdIter>().call2(
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp:419]:
```suggestion
            return hpx::parallel::v1::detail::rotate<FwdIter>().call(
```

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp [hunk 5]
```diff
@@ -308,3 +395,120 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+                "Requires at least forward iterator.");
+
+            typedef std::integral_constant<bool,
+                hpx::is_sequenced_execution_policy<ExPolicy>::value ||
+                    !hpx::traits::is_bidirectional_iterator<FwdIter>::value>
+                is_seq;
+
+            return parallel::util::get_second_element(
+                hpx::parallel::v1::detail::rotate<hpx::parallel::util::in_out_result<FwdIter,FwdIter>>()
+                .call2(std::forward<ExPolicy>(policy), is_seq(), first, new_first, last));
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp:445]: I'm sure our clang-format will complain about this, please always format the files you changed before committing.

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp [hunk 6]
```diff
@@ -308,3 +395,120 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+        friend OutIter tag_fallback_dispatch(
+            hpx::rotate_copy_t, FwdIter first, FwdIter new_first, FwdIter last, 
+                OutIter dest_first)
+        {
+            static_assert((hpx::traits::is_forward_iterator<FwdIter>::value),
+                "Requires at least forward iterator.");
+            static_assert((hpx::traits::is_output_iterator<OutIter>::value),
+                "Requires at least output iterator.");
+
+            //typedef hpx::is_sequenced_execution_policy<ExPolicy> is_seq;
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp:472]: Please remove non-needed comments.

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp [hunk 7]
```diff
@@ -308,3 +395,120 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+
+            return parallel::util::get_second_element(parallel::v1::detail::rotate_copy<
+                hpx::parallel::util::in_out_result<FwdIter, OutIter>>()
+                .call2(hpx::execution::sequenced_policy{}, first,
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp:476]:
```suggestion
                .call(hpx::execution::seq, first, 
```
I believe you will have to add tests for this algorithm overload. Such tests would have caught this.

**Comment by hkaiser** [comment]
Please have an eye on the CircleCI builder (e.g. here for the latest commit: https://app.circleci.com/pipelines/github/STEllAR-GROUP/hpx/8257/workflows/f940a4fb-e466-4a2c-bd70-9127024c4277) as that builder runs a couple of additional forma tests (spellcheck, clang-format, inspect) that need to pass.

**Comment by hcq9102** [comment]
There are some suggestions from Akhil: (Thanks a lot)

>`Inside the HPX_CONCEPT_REQUIRES_ we use hpx::traits::is_iterator instead of hpx::traits::is_forward_iterator or hpx::traits::is_input_iterator.`
---will amend it which I didnt pay more attention.

>`Also everywhere you have used hpx::traits::is_forward_iterator<FwdIter>::value you can change it to hpx::traits::is_forward_iterator_v<FwdIter>,` and  `hpx::traits::range_iterator<Rng>::type, could you change it to hpx::traits::range_iterator_t<Rng>`
only for easy to read
--- will do.

> `wherever we use typedef, could you change it to using?`
--- noticed that but didn't know why, so just use a typedef. will change it today.

**Comment by hcq9102** [comment]
There are some suggestions from Akhil: (Thanks a lot)

`Inside the HPX_CONCEPT_REQUIRES_ we use hpx::traits::is_iterator instead of hpx::traits::is_forward_iterator or hpx::traits::is_input_iterator.`
---didnt pay attention on that before, will amend it.

`Also everywhere you have used hpx::traits::is_forward_iterator<FwdIter>::value you can change it to hpx::traits::is_forward_iterator_v<FwdIter>`, and `hpx::traits::range_iterator<Rng>::type, could you change it to hpx::traits::range_iterator_t<Rng>`
only for easy to read
--- will do.

`wherever we use typedef, could you change it to using?` 
--- noticed that but didn't know why, so just use a typedef. will change it today.

**Comment by hcq9102** [comment]
before doing "uniform format", the commit "add exception and bad alloc tests back" could build_and_test success fully. 
after adding "uniform format", there is one test fail from tests.unit, but it is not related to my changed files.

so I think this PR is ready for review.

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp [hunk 8]
```diff
@@ -236,75 +368,145 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+        friend typename parallel::util::detail::algorithm_result<ExPolicy,
+            FwdIter>::type
+        tag_fallback_dispatch(hpx::rotate_t, ExPolicy&& policy, FwdIter first,
+            FwdIter new_first, FwdIter last)
+        {
+            static_assert((hpx::traits::is_forward_iterator_v<FwdIter>),
+                "Requires at least forward iterator.");
+
+            using is_seq = std::integral_constant<bool,
+                hpx::is_sequenced_execution_policy<ExPolicy>::value ||
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp:441]: FWIW, we have a `hpx::is_sequenced_execution_policy_v<ExPolicy>` as well (https://github.com/STEllAR-GROUP/hpx/blob/master/libs/parallelism/execution/include/hpx/execution/traits/is_execution_policy.hpp#L116-L118).

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp [hunk 9]
```diff
@@ -122,26 +363,291 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+#include <utility>
+
+namespace hpx { namespace parallel { inline namespace v1 {
+
+    // clang-format off
+    template <typename ExPolicy, typename Rng,
+        HPX_CONCEPT_REQUIRES_(
+            hpx::is_execution_policy<ExPolicy>::value &&
+            hpx::traits::is_range<Rng>::value
+        )>
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp:409]: `hpx::is_execution_policy_v<ExPolicy>` is defined here: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/parallelism/execution/include/hpx/execution/traits/is_execution_policy.hpp#L67-L69, we still miss a `hpx::traits::is_range_v<Rng>` however (see: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/iterator_support/include/hpx/iterator_support/traits/is_range.hpp)

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp [hunk 10]
```diff
@@ -122,26 +363,291 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
+            FwdIter1 first, FwdIter1 middle, Sent last, FwdIter2 dest_first)
+        {
+            static_assert((hpx::traits::is_forward_iterator_v<FwdIter1>),
+                "Requires at least forward iterator.");
+            static_assert((hpx::traits::is_forward_iterator_v<FwdIter2>),
+                "Requires at least forward iterator.");
+
+            using is_seq = std::integral_constant<bool,
+                hpx::is_sequenced_execution_policy<ExPolicy>::value ||
+                    !hpx::traits::is_forward_iterator_v<FwdIter1>>;
```
Comment (hkaiser) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp:598]: I don't think this is correct. A few lines above you verify that the iterator is a `forward_iterator`. I think the correct way would be:
```
        using is_seq = std::integral_constant<bool,
                hpx::is_sequenced_execution_policy<ExPolicy>::value ||
                    !hpx::traits::is_bidirectional_iterator_v<FwdIter1>>;
```

**Review by hkaiser — DISMISSED** [review]
Except for the minor comments, this LGTM, thanks a lot! Good work!

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp (hunk 8 above)
Comment (hcq9102) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp:441]: Got it, will check and replace it.

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp (hunk 9 above)
Comment (hcq9102) [reply libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp:409]: ok, I see.  I will check and see if I can add a `hpx::traits::is_range_v<Rng>` for `hpx::traits::is_range<Rng>::value` later, I mean as another minor task. if you think this is necessary for our future work.

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp (hunk 10 above)
Comment (hcq9102) [reply libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp:598]: Sorry for that, will amend it.

File: libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp (hunk 8 above)
Comment (hcq9102) [reply libs/parallelism/algorithms/include/hpx/parallel/algorithms/rotate.hpp:441]: Got it, will check and replace it.

**Comment by hcq9102** [comment]
> Except for the minor comments, this LGTM, thanks a lot! Good work!

Thanks, Appreciate your help!

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp (hunk 9 above)
Comment (hkaiser) [reply libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/rotate.hpp:409]: We have a lot of similar things to work on. This is nothing for this PR, though.

**Comment by hcq9102** [comment]
two more commits, finally, the **build-and-test** passed sucessfully. 
But just find pushing new commits dismissed your approval, sorry about that.  @hkaiser

**Comment by hkaiser** [comment]
> two more commits, finally, the **build-and-test** passed sucessfully.
> But just find pushing new commits dismissed your approval, sorry about that. @hkaiser
> won't happen.

Let's get this merged asap, then.

**Comment by hkaiser** [comment]
> But just find pushing new commits dismissed your approval, sorry about that. @hkaiser
> won't happen.

There is nothing you can do about this - it's how github handles PRs.

**Comment by hcq9102** [comment]
> > But just find pushing new commits dismissed your approval, sorry about that. @hkaiser
> > won't happen.
> 
> There is nothing you can do about this - it's how github handles PRs.

got it.

**Comment by hkaiser** [comment]
Congratulations, your first PR to the HPX repository has just been merged! As a 'thank you' we offer a free STE||AR-Group t-shirt to all of our first time contributors. If you are interested in receiving one, please get back to me directly so we can set up the delivery.

**Comment by hcq9102** [comment]
> Congratulations, your first PR to the HPX repository has just been merged! As a 'thank you' we offer a free STE||AR-Group t-shirt to all of our first time contributors. If you are interested in receiving one, please get back to me directly so we can set up the delivery.

Thanks a lot!!

**Pull request merged**